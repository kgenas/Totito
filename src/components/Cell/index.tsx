import './Cell.css';
import classNames from 'classnames';
import { useSelector, useDispatch } from 'react-redux';
import { AppDispatch, RootState } from '../../app/store';
import { winnerCalculatorValue } from '../../features/game/result';
import { click } from '../../features/game/totito';

interface valueProps {    
    key: number ;
    value: string;
    valueKey: number;
    switchOnOff: boolean;
}
export const Cell = ( props: valueProps ) => {
          
    const { cellValues, numberOfTurn, xIsNext } = useSelector((state: RootState) => state.game );    
    const dispatch:AppDispatch = useDispatch();

    const isCellEmpty = ( cellIndex:number ) => cellValues[cellIndex] === '';
    const onCellClicked = () =>{ 
      
      if (isCellEmpty( props.valueKey )){
        const newCellValues = [...cellValues];
        newCellValues[props.valueKey] = xIsNext ? 'X':'O';
        dispatch(winnerCalculatorValue( newCellValues,(numberOfTurn-1),props.valueKey ));        
        dispatch(click(
            {
                cellValues: newCellValues,
                numberOfTurn: (numberOfTurn-1),
                xIsNext: !xIsNext,
            }
        ));
    }
    }
    const cellCss = classNames({
        cell: true,
        winner: props.switchOnOff
    })

    const cellContentCss = classNames({
        'cell--contet': true,
        populated: props.value 
    })

  return (
    <button
        className={ cellCss }
        onClick = { onCellClicked }
    >
        <span
            className={ cellContentCss }
        >
            { props.value }
        </span>
    </button>
  )
}
