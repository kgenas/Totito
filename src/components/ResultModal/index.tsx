import classNames from 'classnames';
import { useSelector, useDispatch } from 'react-redux';
import { newGame } from '../../features/game/totito';
import './ResultModal.css';
import { AppDispatch, RootState } from '../../app/store';
import { winnerCalculatorStart } from '../../features/game/result';

export const ResultModal = () =>{

    const dispatch:AppDispatch = useDispatch();
    const { winner , isGameOver } = useSelector((state: RootState) => state.result );    

    const resultMondalClasses = classNames({
        'modal-overlay': true,
        'modal-overlay--modal-open': isGameOver

    })
    const message = winner ? `Winner is ${ winner }.` :'It is a tie';

    const newGames = () =>{
        dispatch(newGame());
        dispatch(winnerCalculatorStart());
    }
    return(
        <div className= { resultMondalClasses }>
            <div className='game-result-modal'>
                <div className='result-container'>
                    <div className='winner-container'>
                        <span>{ message }</span>
                    </div>
                </div>
                <div className='new-game-container'>
                    <button
                        className='new-game-container__button'
                        onClick={ newGames }
                    >
                        Start New Game
                    </button>
                </div>
            </div>            
        </div>
    )
}