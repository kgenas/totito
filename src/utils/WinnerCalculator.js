const winnningMatrix = {
    0:[[1,2],[3,6],[4,8]], //? example 0,1,2 o 0,3,6 o 0,4,8
    1:[[0,2],[4,7]],
    2:[[0,1],[5,8],[4,6]],
    3:[[0,6],[4,5]],
    4:[[2,6],[3,5],[1,7],[0,8]],
    5:[[3,4],[2,8]],
    6:[[7,8],[0,3],[2,4]],
    7:[[6,8],[1,4]],
    8:[[6,7],[2,5],[0,4]]
}

export const WinnerCalculator = (cellValues, numberOfTurn, cellIndex) =>{

    const WinningOptions = winnningMatrix[cellIndex];
    for(let WinningOption of WinningOptions){
        
        const currentValue = cellValues[cellIndex];
        const firstOption =  cellValues[WinningOption[0]];
        const secondOption = cellValues[WinningOption[1]];

        if(currentValue === firstOption && firstOption === secondOption ){ 
            return{           
            finish:true,
            winner:currentValue,
            winningCombination:[cellIndex,WinningOption[0],WinningOption[1]]
        };
        }
    }

    if( numberOfTurn === 0 ){
        return{
            finish: true,
            winner: undefined,
            winningCombination:[]
        }
    }
    return{
        finish: false,
        winner: undefined,
        winningCombination:[]
    }
}