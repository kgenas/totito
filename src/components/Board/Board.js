import './Board.css';
import { useSelector } from 'react-redux';
import { Cell } from '../Cell/Cell';

export const Board = (props) => {
  
  const { cellValues, winningCombination } = useSelector( state => state.Game)
  const cell = cellValues.map((value, index)=>{
        
    const switchOnOff = winningCombination && winningCombination.indexOf(index) >= 0;
    return(
        <Cell
            key= { index }
            value = { value }
            switchOnOff = { switchOnOff }
            onClick = { () => props.cellClicked(index) }
        />
    )
  })

  return (
    <div className='board'>
        { cell }
    </div>
  )
}
