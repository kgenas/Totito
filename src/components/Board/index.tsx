import './Board.css';
import { useSelector } from 'react-redux';
import { Cell } from '../Cell';
import { RootState } from '../../app/store';
export const Board = () => {
    
  const { cellValues }  = useSelector((state: RootState) => state.game );
  const { winningCombination } = useSelector((state: RootState) => state.result ); 

  const cell = cellValues.map((value:string, index:number)=>{
       
  const switchOnOff = winningCombination && winningCombination.indexOf( index ) >= 0;
    return(
        <Cell
            key = { index }
            value = { value }
            valueKey = { index }
            switchOnOff = { switchOnOff }
        />
    )
  })

  return (
    <div className='board'>
        { cell }
    </div>
  )
}
