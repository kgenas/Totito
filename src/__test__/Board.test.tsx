/* eslint-disable testing-library/no-debugging-utils */
import React from 'react';
import { screen } from '@testing-library/react';
import { Board } from '../components/Board';
import { renderWithProviders } from '../utils/test-utils';


describe('Board',() =>{
    // Note: running cleanup afterEach is done automatically for you in @testing-library/react@9.0.0 or higher
    // unmount and cleanup DOM after the test is finished.
    //? afterEach(cleanup);
    test('Renders Board component',() => {
        renderWithProviders(<Board/>);
        screen.debug(); 
    });
    
    test('Matches snapshot', () => { 
        const { baseElement } = renderWithProviders(<Board/>);
        expect( baseElement ).toMatchSnapshot();
    })    
});
