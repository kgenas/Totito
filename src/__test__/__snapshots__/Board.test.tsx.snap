// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Matches snapshot 1`] = `
<body>
  <div>
    <div
      class="board"
    >
      <button
        class="cell"
      >
        <span
          class="cell--contet"
        />
      </button>
      <button
        class="cell"
      >
        <span
          class="cell--contet"
        />
      </button>
      <button
        class="cell"
      >
        <span
          class="cell--contet"
        />
      </button>
      <button
        class="cell"
      >
        <span
          class="cell--contet"
        />
      </button>
      <button
        class="cell"
      >
        <span
          class="cell--contet"
        />
      </button>
      <button
        class="cell"
      >
        <span
          class="cell--contet"
        />
      </button>
      <button
        class="cell"
      >
        <span
          class="cell--contet"
        />
      </button>
      <button
        class="cell"
      >
        <span
          class="cell--contet"
        />
      </button>
      <button
        class="cell"
      >
        <span
          class="cell--contet"
        />
      </button>
    </div>
  </div>
</body>
`;
